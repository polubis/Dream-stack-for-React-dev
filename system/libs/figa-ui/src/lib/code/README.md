```ts
import { EditorView } from '@codemirror/view';
import type { Extension } from '@codemirror/state';
import { HighlightStyle, syntaxHighlighting } from '@codemirror/language';
import { tags as t } from '@lezer/highlight';
import { tokens } from '../theme-provider';

const CODE_LINE_HEIGHT = 22;

const dark = {
  editor: EditorView.theme({
    '&': {
      backgroundColor: '#050404',
    },

    // '.cm-content': {
    //   caretColor: 'red',
    //   padding: 0,
    // },

    '.cm-scroller': {
      fontSize: '1.6rem',
    },

    // '.selectionLayer': {
    //   backgroundColor: 'red',
    // },

    '.cm-cursor, .cm-dropCursor': { borderLeftColor: '#fff' },
    '&.cm-focused .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection':
      { backgroundColor: '#404040' },

    // '.cm-panels': { backgroundColor: darkBackground, color: fg1 },
    // '.cm-panels.cm-panels-top': { borderBottom: '2px solid black' },
    // '.cm-panels.cm-panels-bottom': { borderTop: '2px solid black' },

    // '.cm-searchMatch': {
    //   backgroundColor: bg0,
    //   color: yellow,
    //   outline: `1px solid ${bg3}`,
    // },
    // '.cm-searchMatch.cm-searchMatch-selected': {
    //   backgroundColor: bg3,
    // },

    // '.cm-activeLine': { backgroundColor: highlightBackground },
    // '.cm-selectionMatch': { backgroundColor: bg3 },

    // '&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket': {
    //   outline: `1px solid ${bg3}`,
    //   fontStyle: 'bold',
    // },

    // '&.cm-focused .cm-matchingBracket': {
    //   backgroundColor: bg3,
    // },

    '*[title="Fold line"], *[title="Unfold line"]': {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: tokens.spacing[300],
    },

    '.cm-gutters': {
      backgroundColor: '#050404',
      color: '#fff',
      border: 'none',
    },

    '.cm-activeLineGutter': {
      backgroundColor: '#4d4d4d',
    },

    // '.cm-foldPlaceholder': {
    //   backgroundColor: 'transparent',
    //   border: 'none',
    //   color: '#ddd',
    // },

    // '.cm-tooltip': {
    //   border: 'none',
    //   backgroundColor: tooltipBackground,
    // },
    // '.cm-tooltip .cm-tooltip-arrow:before': {
    //   borderTopColor: 'transparent',
    //   borderBottomColor: 'transparent',
    // },
    // '.cm-tooltip .cm-tooltip-arrow:after': {
    //   borderTopColor: tooltipBackground,
    //   borderBottomColor: tooltipBackground,
    // },
    // '.cm-tooltip-autocomplete': {
    //   '& > ul > li[aria-selected]': {
    //     backgroundColor: highlightBackground,
    //     color: fg2,
    //   },
    // },
  }),
  content: HighlightStyle.define([
    // { tag: t.angleBracket, color: 'red' },
    // { tag: t.annotation, color: 'red' },
    { tag: t.arithmeticOperator, color: '#fff' },
    // { tag: t.atom, color: 'red' },
    // { tag: t.attributeName, color: 'red' },
    // { tag: t.attributeValue, color: 'red' },
    { tag: t.bitwiseOperator, color: '#fff' },
    { tag: t.blockComment, color: '#6A9955' },
    { tag: t.bool, color: '#569cd6' },
    { tag: t.brace, color: '#ffd70f' },
    { tag: t.bracket, color: '#da6bad' },
    // { tag: t.changed, color: 'red' },
    // { tag: t.character, color: 'red' },
    { tag: t.className, color: '#3abfb0' },
    // { tag: t.color, color: 'red' },
    { tag: t.comment, color: '#6A9955' },
    { tag: t.compareOperator, color: '#fff' },
    // { tag: t.constant(t.name), color: 'red' },
    // { tag: t.content, color: 'red' },
    // { tag: t.contentSeparator, color: 'red' },
    { tag: t.controlKeyword, color: '#eba0d9' },
    // { tag: t.controlOperator, color: 'red' },
    { tag: t.definition(t.name), color: '#4fc1ff' },
    { tag: t.definitionKeyword, color: '#3981c5' },
    { tag: t.definitionOperator, color: '#fff' },
    // { tag: t.deleted, color: 'red' },
    { tag: t.derefOperator, color: '#fff' },
    // { tag: t.docComment, color: 'red' },
    // { tag: t.docString, color: 'red' },
    // { tag: t.documentMeta, color: 'red' },
    // { tag: t.emphasis, color: 'red' },
    // { tag: t.escape, color: 'red' },
    // { tag: t.float, color: 'red' },
    { tag: t.function(t.name), color: '#dcdc9f' },
    // { tag: t.heading, color: 'red' },
    // { tag: t.heading1, color: 'red' },
    // { tag: t.heading2, color: 'red' },
    // { tag: t.heading3, color: 'red' },
    // { tag: t.heading4, color: 'red' },
    // { tag: t.heading5, color: 'red' },
    // { tag: t.heading6, color: 'red' },
    // { tag: t.inserted, color: 'red' },
    // { tag: t.integer, color: 'red' },
    // { tag: t.invalid, color: 'red' },
    { tag: t.keyword, color: '#3981c5' },
    { tag: t.labelName, color: '#4fc1ff' },
    { tag: t.lineComment, color: '#6A9955' },
    // { tag: t.link, color: 'red' },
    // { tag: t.list, color: 'red' },
    { tag: t.literal, color: '#ef8c59' },
    { tag: t.local(t.name), color: 'red' },
    { tag: t.logicOperator, color: '#fff' },
    // { tag: t.macroName, color: 'red' },
    // { tag: t.meta, color: 'red' },
    // { tag: t.modifier, color: 'red' },
    { tag: t.moduleKeyword, color: '#eba0d9' },
    // { tag: t.monospace, color: 'red' },
    { tag: t.name, color: '#4fc1ff' },
    // { tag: t.namespace, color: 'red' },
    { tag: t.null, color: '#569cd6' },
    { tag: t.number, color: '#accea8' },
    { tag: t.operator, color: 'red' },
    { tag: t.operatorKeyword, color: '#3abfb0' },
    { tag: t.paren, color: '#da6bad' },
    // { tag: t.processingInstruction, color: 'red' },
    // { tag: t.propertyName, color: 'red' },
    // { tag: t.punctuation, color: 'red' },
    // { tag: t.quote, color: 'red' },
    // { tag: t.regexp, color: 'red' },
    // { tag: t.self, color: 'red' },
    // { tag: t.separator, color: 'red' },
    // { tag: t.special(t.name), color: 'red' },
    // { tag: t.squareBracket, color: 'red' },
    // { tag: t.standard(t.name), color: 'red' },
    // { tag: t.strikethrough, color: 'red' },
    // { tag: t.string, color: 'red' },
    // { tag: t.strong, color: 'red' },
    // { tag: t.tagName, color: 'red' },
    { tag: t.typeName, color: '#3abfb0' },
    // { tag: t.typeOperator, color: 'red' },
    // { tag: t.unit, color: 'red' },
    // { tag: t.updateOperator, color: 'red' },
    // { tag: t.url, color: 'red' },
    // { tag: t.variableName, color: 'red' },
  ]),
};

const DEFAULT_THEME: Extension = [
  dark.editor,
  syntaxHighlighting(dark.content),
];

export { CODE_LINE_HEIGHT, DEFAULT_THEME };
```