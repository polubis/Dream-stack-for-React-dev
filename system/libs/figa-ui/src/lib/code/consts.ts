import { EditorView } from '@codemirror/view';
import type { Extension } from '@codemirror/state';
import { HighlightStyle, syntaxHighlighting } from '@codemirror/language';
import { tags as t } from '@lezer/highlight';
import { tokens } from '../theme-provider';

const CODE_LINE_HEIGHT = 22;

const dark0 = '#282828',
  dark1 = '#3c3836',
  dark3 = '#665c54',
  gray_245 = '#928374',
  light1 = '#ebdbb2',
  light2 = '#d5c4a1',
  light3 = '#bdae93',
  light4 = '#a89984',
  bright_red = '#fb4934',
  bright_green = '#b8bb26',
  bright_yellow = '#fabd2f',
  bright_blue = '#83a598',
  bright_purple = '#d3869b',
  bright_aqua = '#8ec07c',
  bright_orange = '#fe8019';

const bg0 = dark0,
  bg1 = dark1,
  bg3 = dark3,
  gray = gray_245,
  fg1 = light1,
  fg2 = light2,
  fg3 = light3,
  fg4 = light4,
  red = bright_red,
  green = bright_green,
  yellow = bright_yellow,
  blue = bright_blue,
  purple = bright_purple,
  aqua = bright_aqua,
  orange = bright_orange;

const invalid = red,
  darkBackground = bg1,
  highlightBackground = darkBackground,
  background = bg0,
  tooltipBackground = bg1,
  selection = darkBackground,
  cursor = orange;

const gruvboxDarkTheme = EditorView.theme(
  {
    '&': {
      color: 'red',
      backgroundColor: background,
    },

    '.cm-content': {
      caretColor: cursor,
      padding: 0,
    },

    '.cm-scroller': {
      fontSize: '1.6rem',
    },

    '.cm-cursor, .cm-dropCursor': { borderLeftColor: cursor },
    '&.cm-focused .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection':
      { backgroundColor: selection },

    '.cm-panels': { backgroundColor: darkBackground, color: fg1 },
    '.cm-panels.cm-panels-top': { borderBottom: '2px solid black' },
    '.cm-panels.cm-panels-bottom': { borderTop: '2px solid black' },

    '.cm-searchMatch': {
      backgroundColor: bg0,
      color: yellow,
      outline: `1px solid ${bg3}`,
    },
    '.cm-searchMatch.cm-searchMatch-selected': {
      backgroundColor: bg3,
    },

    '.cm-activeLine': { backgroundColor: highlightBackground },
    '.cm-selectionMatch': { backgroundColor: bg3 },

    '&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket': {
      outline: `1px solid ${bg3}`,
      fontStyle: 'bold',
    },

    '&.cm-focused .cm-matchingBracket': {
      backgroundColor: bg3,
    },

    '*[title="Fold line"], *[title="Unfold line"]': {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: tokens.spacing[300]
    },

    '.cm-gutters': {
      backgroundColor: '#282a36',
      color: '#fff',
      border: 'none',
    },

    '.cm-activeLineGutter': {
      backgroundColor: highlightBackground,
    },

    '.cm-foldPlaceholder': {
      backgroundColor: 'transparent',
      border: 'none',
      color: '#ddd',
    },

    '.cm-tooltip': {
      border: 'none',
      backgroundColor: tooltipBackground,
    },
    '.cm-tooltip .cm-tooltip-arrow:before': {
      borderTopColor: 'transparent',
      borderBottomColor: 'transparent',
    },
    '.cm-tooltip .cm-tooltip-arrow:after': {
      borderTopColor: tooltipBackground,
      borderBottomColor: tooltipBackground,
    },
    '.cm-tooltip-autocomplete': {
      '& > ul > li[aria-selected]': {
        backgroundColor: highlightBackground,
        color: fg2,
      },
    },
  },
  { dark: true }
);

const gruvboxDarkHighlightStyle = HighlightStyle.define([
  { tag: t.keyword, color: red },
  {
    tag: [t.name, t.deleted, t.character, t.propertyName, t.macroName],
    color: aqua,
  },
  { tag: [t.variableName], color: blue },
  { tag: [t.function(t.variableName)], color: green, fontStyle: 'bold' },
  { tag: [t.labelName], color: fg1 },
  {
    tag: [t.color, t.constant(t.name), t.standard(t.name)],
    color: purple,
  },
  { tag: [t.definition(t.name), t.separator], color: fg1 },
  { tag: [t.brace], color: fg1 },
  {
    tag: [t.annotation],
    color: invalid,
  },
  {
    tag: [t.number, t.changed, t.annotation, t.modifier, t.self, t.namespace],
    color: purple,
  },
  {
    tag: [t.typeName, t.className],
    color: yellow,
  },
  {
    tag: [t.operator, t.operatorKeyword],
    color: red,
  },
  {
    tag: [t.tagName],
    color: aqua,
    fontStyle: 'bold',
  },
  {
    tag: [t.squareBracket],
    color: orange,
  },
  {
    tag: [t.angleBracket],
    color: blue,
  },
  {
    tag: [t.attributeName],
    color: aqua,
  },
  {
    tag: [t.regexp],
    color: aqua,
  },
  {
    tag: [t.quote],
    color: gray,
  },
  { tag: [t.string], color: fg1 },
  {
    tag: t.link,
    color: fg4,
    textDecoration: 'underline',
    textUnderlinePosition: 'under',
  },
  {
    tag: [t.url, t.escape, t.special(t.string)],
    color: purple,
  },
  { tag: [t.meta], color: yellow },
  { tag: [t.comment], color: gray, fontStyle: 'italic' },
  { tag: t.strong, fontWeight: 'bold', color: orange },
  { tag: t.emphasis, fontStyle: 'italic', color: green },
  { tag: t.strikethrough, textDecoration: 'line-through' },
  { tag: t.heading, fontWeight: 'bold', color: green },
  { tag: [t.heading1, t.heading2], fontWeight: 'bold', color: green },
  {
    tag: [t.heading3, t.heading4],
    fontWeight: 'bold',
    color: yellow,
  },
  {
    tag: [t.heading5, t.heading6],
    color: yellow,
  },
  { tag: [t.atom, t.bool, t.special(t.variableName)], color: purple },
  {
    tag: [t.processingInstruction, t.inserted],
    color: bright_blue,
  },
  {
    tag: [t.contentSeparator],
    color: red,
  },
  { tag: t.invalid, color: orange, borderBottom: `1px dotted ${invalid}` },
]);

const DEFAULT_THEME: Extension = [
  gruvboxDarkTheme,
  syntaxHighlighting(gruvboxDarkHighlightStyle),
];

export { CODE_LINE_HEIGHT, DEFAULT_THEME };
