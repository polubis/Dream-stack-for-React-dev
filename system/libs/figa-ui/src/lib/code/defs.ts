import type { Theme } from '../theme-provider';

type CodeLang = 'js' | 'md';

interface CodeProps {
  className?: string;
  children: string;
  readonly?: boolean;
  lang?: CodeLang;
  wrapLines?: boolean;
  onChange?: (value: string) => void;
}

interface SetupConfig {
  children: CodeProps['children'];
  readonly: CodeProps['readonly'];
  lang: CodeLang;
  wrapLines: CodeProps['wrapLines'];
  parent: Element;
  theme: Theme['code'];
  onChange?: CodeProps['onChange'];
}

export type { CodeProps, SetupConfig, CodeLang };
