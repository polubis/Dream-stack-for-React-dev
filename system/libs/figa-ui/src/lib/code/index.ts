export * from './code';
export * from './defs';
