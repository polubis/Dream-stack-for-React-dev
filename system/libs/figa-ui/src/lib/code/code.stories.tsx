import type { Story, Meta } from '@storybook/react';
import type { CodeProps } from './defs';

import { Code } from './code';

export default {
  component: Code,
  title: 'Code',
} as Meta;

const Template: Story<CodeProps> = (props) => <Code {...props} />;

const children = `
const { mergeConfig } = require('vite');
const viteTsConfigPaths = require('vite-tsconfig-paths').default;

module.exports = {
  core: { builder: '@storybook/builder-vite' },
  stories: [
    '../src/lib/**/*.stories.mdx',
    '../src/lib/**/*.stories.@(js|jsx|ts|tsx)',
  ],
  addons: ['@storybook/addon-essentials'],
  staticDirs: [{ from: '../../../assets', to: '/' }],
  async viteFinal(config, { configType }) {
    return mergeConfig(config, {
      plugins: [
        viteTsConfigPaths({
          root: '../../../',
        }),
      ],
    });
  },
};

// To customize your Vite configuration you can use the viteFinal field.
// Check https://storybook.js.org/docs/react/builders/vite#configuration
// and https://nx.dev/packages/storybook/documents/custom-builder-configs
`;

export const Default = Template.bind({});
Default.args = {
  children,
};

export const Readonly = Template.bind({});
Readonly.args = {
  children,
  readonly: true,
};
