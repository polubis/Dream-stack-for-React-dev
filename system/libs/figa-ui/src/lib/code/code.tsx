import type { CodeProps } from './defs';

import { basicSetup } from 'codemirror';
import { EditorView } from '@codemirror/view';
import { javascript } from '@codemirror/lang-javascript';
import { useLayoutEffect, useRef } from 'react';
import { CODE_LINE_HEIGHT, DEFAULT_THEME } from './consts';
import c from 'classnames';

const CodeContent = ({ children, className }: CodeProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    const parent = ref.current;

    if (!parent) return;

    const view = new EditorView({
      doc: children,
      extensions: [basicSetup, javascript(), DEFAULT_THEME],
      parent,
    });

    return () => {
      view.destroy();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return <div className={c('code', className)} ref={ref} />;
};

const Code = ({ children, className }: CodeProps) => {
  if (typeof children !== 'string') {
    throw Error('Children must be a string');
  }

  if (typeof window === 'undefined') {
    return (
      <div
        className={c('code', className)}
        style={{ height: children.split('\n').length * CODE_LINE_HEIGHT }}
      />
    );
  }

  return <CodeContent children={children} className={className} />;
};

export { Code };
