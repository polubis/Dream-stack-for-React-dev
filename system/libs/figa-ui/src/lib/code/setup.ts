import { basicSetup } from 'codemirror';
import { EditorView } from '@codemirror/view';
import { javascript } from '@codemirror/lang-javascript';
import { DEFAULT_THEME } from './consts';
import { EditorState, type Extension } from '@codemirror/state';
import type { SetupConfig } from './defs';

const selectionExtensionIdx = 15;

export const setup = ({
  readonly,
  children,
  parent,
  onChange,
}: SetupConfig): EditorView => {
  let setup = basicSetup;
  const extensions: Extension[] = [javascript(), DEFAULT_THEME];

  if (readonly) {
    setup = (setup as Extension[]).filter(
      (_, i) => i !== selectionExtensionIdx
    );
    extensions.push(EditorState.readOnly.of(true));
  }

  extensions.push(setup);

  const view = new EditorView({
    doc: children,
    extensions: [
      ...extensions,
      ...(!readonly && onChange
        ? [
            EditorView.updateListener.of((v) => {
              if (v.docChanged) {
                onChange(view.state.doc.toString());
              }
            }),
          ]
        : []),
    ],
    parent,
  });

  return view;
};
