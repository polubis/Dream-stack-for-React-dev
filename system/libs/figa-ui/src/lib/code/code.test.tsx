import { render, screen } from '@testing-library/react';
import { Code } from './code';

describe('Code can be used when: ', () => {
  const children = `  const { mergeConfig } = require('vite');
  const viteTsConfigPaths = require('vite-tsconfig-paths').default;
  
  module.exports = {
    core: { builder: '@storybook/builder-vite' },
    stories: [
      '../src/lib/**/*.stories.mdx',
      '../src/lib/**/*.stories.@(js|jsx|ts|tsx)',
    ],
    addons: ['@storybook/addon-essentials'],
    staticDirs: [{ from: '../../../assets', to: '/' }],
    async viteFinal(config, { configType }) {
      return mergeConfig(config, {
        plugins: [
          viteTsConfigPaths({
            root: '../../../',
          }),
        ],
      });
    },
  };
  
  // To customize your Vite configuration you can use the viteFinal field.
  // Check https://storybook.js.org/docs/react/builders/vite#configuration
  // and https://nx.dev/packages/storybook/documents/custom-builder-configs`;

  it('renders given code snippet', () => {
    render(<Code children={children} />);

    screen.getByText(
      /To customize your Vite configuration you can use the viteFinal field./
    );
  });

  it('[FRAGILE] assigns class names for code component', () => {
    const { container } = render(
      <Code children={children} className="my-class" />
    );

    expect(container.querySelector('.my-class')).toBeTruthy();
  });

  describe('during readonly mode', () => {
    it('[FRAGILE] disables selection', () => {
      const readonly = render(<Code children={children} readonly />);

      expect(readonly.container.querySelector('.cm-activeLine')).toBeFalsy();

      const editable = render(<Code children={children} />);

      expect(editable.container.querySelector('.cm-activeLine')).toBeTruthy();
    });
  });
});
