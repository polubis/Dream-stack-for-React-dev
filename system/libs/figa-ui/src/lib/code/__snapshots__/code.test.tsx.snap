// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code can be used when:  assigns class names for code component 1`] = `
<DocumentFragment>
  <div
    class="code my-class"
  >
    <div
      class="cm-editor ͼ1 ͼ3 ͼ4 ͼo"
    >
      <div
        aria-live="polite"
        style="position: fixed; top: -10000px;"
      />
      <div
        class="cm-scroller"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="cm-gutters"
          style="min-height: 350px; position: sticky;"
        >
          <div
            class="cm-gutter cm-lineNumbers"
          >
            <div
              class="cm-gutterElement"
              style="height: 0px; visibility: hidden; pointer-events: none;"
            >
              99
            </div>
            <div
              class="cm-gutterElement cm-activeLineGutter"
              style="height: 14px;"
            >
              1
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              2
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              3
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              4
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              5
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              6
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              7
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              8
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              9
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              10
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              11
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              12
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              13
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              14
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              15
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              16
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              17
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              18
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              19
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              20
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              21
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              22
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              23
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              24
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              25
            </div>
          </div>
          <div
            class="cm-gutter cm-foldGutter"
          >
            <div
              class="cm-gutterElement"
              style="height: 0px; visibility: hidden; pointer-events: none;"
            >
              <span
                title="Unfold line"
              >
                ›
              </span>
            </div>
            <div
              class="cm-gutterElement cm-activeLineGutter"
              style="height: 14px;"
            />
            <div
              class="cm-gutterElement"
              style="height: 14px; margin-top: 28px;"
            >
              <span
                title="Fold line"
              >
                ⌄
              </span>
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px; margin-top: 14px;"
            >
              <span
                title="Fold line"
              >
                ⌄
              </span>
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px; margin-top: 70px;"
            >
              <span
                title="Fold line"
              >
                ⌄
              </span>
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              <span
                title="Fold line"
              >
                ⌄
              </span>
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              <span
                title="Fold line"
              >
                ⌄
              </span>
            </div>
            <div
              class="cm-gutterElement"
              style="height: 14px;"
            >
              <span
                title="Fold line"
              >
                ⌄
              </span>
            </div>
          </div>
        </div>
        <div
          aria-autocomplete="list"
          aria-multiline="true"
          autocapitalize="off"
          autocorrect="off"
          class="cm-content"
          contenteditable="true"
          data-language="javascript"
          role="textbox"
          spellcheck="false"
          style="tab-size: 4"
          translate="no"
        >
          <div
            class="cm-activeLine cm-line"
          >
              
            <span
              class="ͼp"
            >
              const
            </span>
             
            <span
              class="ͼw"
            >
              {
            </span>
             
            <span
              class="ͼq"
            >
              mergeConfig
            </span>
             
            <span
              class="ͼw"
            >
              }
            </span>
             
            <span
              class="ͼ10"
            >
              =
            </span>
             
            <span
              class="ͼs"
            >
              require
            </span>
            (
            <span
              class="ͼ17"
            >
              'vite'
            </span>
            )
            <span
              class="ͼv"
            >
              ;
            </span>
          </div>
          <div
            class="cm-line"
          >
              
            <span
              class="ͼp"
            >
              const
            </span>
             
            <span
              class="ͼr"
            >
              viteTsConfigPaths
            </span>
             
            <span
              class="ͼ10"
            >
              =
            </span>
             
            <span
              class="ͼs"
            >
              require
            </span>
            (
            <span
              class="ͼ17"
            >
              'vite-tsconfig-paths'
            </span>
            )
            <span
              class="ͼ10"
            >
              .
            </span>
            <span
              class="ͼq"
            >
              default
            </span>
            <span
              class="ͼv"
            >
              ;
            </span>
          </div>
          <div
            class="cm-line"
          >
              
          </div>
          <div
            class="cm-line"
          >
              
            <span
              class="ͼr"
            >
              module
            </span>
            <span
              class="ͼ10"
            >
              .
            </span>
            <span
              class="ͼq"
            >
              exports
            </span>
             
            <span
              class="ͼ10"
            >
              =
            </span>
             
            <span
              class="ͼw"
            >
              {
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼq"
            >
              core
            </span>
            : 
            <span
              class="ͼw"
            >
              {
            </span>
             
            <span
              class="ͼq"
            >
              builder
            </span>
            : 
            <span
              class="ͼ17"
            >
              '@storybook/builder-vite'
            </span>
             
            <span
              class="ͼw"
            >
              }
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼq"
            >
              stories
            </span>
            : 
            <span
              class="ͼ12"
            >
              [
            </span>
          </div>
          <div
            class="cm-line"
          >
                  
            <span
              class="ͼ17"
            >
              '../src/lib/**/*.stories.mdx'
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                  
            <span
              class="ͼ17"
            >
              '../src/lib/**/*.stories.@(js|jsx|ts|tsx)'
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼ12"
            >
              ]
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼq"
            >
              addons
            </span>
            : 
            <span
              class="ͼ12"
            >
              [
            </span>
            <span
              class="ͼ17"
            >
              '@storybook/addon-essentials'
            </span>
            <span
              class="ͼ12"
            >
              ]
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼq"
            >
              staticDirs
            </span>
            : 
            <span
              class="ͼ12"
            >
              [
            </span>
            <span
              class="ͼw"
            >
              {
            </span>
             
            <span
              class="ͼq"
            >
              from
            </span>
            : 
            <span
              class="ͼ17"
            >
              '../../../assets'
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
             
            <span
              class="ͼq"
            >
              to
            </span>
            : 
            <span
              class="ͼ17"
            >
              '/'
            </span>
             
            <span
              class="ͼw"
            >
              }
            </span>
            <span
              class="ͼ12"
            >
              ]
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼy"
            >
              async
            </span>
             
            <span
              class="ͼq"
            >
              viteFinal
            </span>
            (
            <span
              class="ͼr"
            >
              config
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
             
            <span
              class="ͼw"
            >
              {
            </span>
             
            <span
              class="ͼq"
            >
              configType
            </span>
             
            <span
              class="ͼw"
            >
              }
            </span>
            ) 
            <span
              class="ͼw"
            >
              {
            </span>
          </div>
          <div
            class="cm-line"
          >
                  
            <span
              class="ͼp"
            >
              return
            </span>
             
            <span
              class="ͼs"
            >
              mergeConfig
            </span>
            (
            <span
              class="ͼr"
            >
              config
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
             
            <span
              class="ͼw"
            >
              {
            </span>
          </div>
          <div
            class="cm-line"
          >
                    
            <span
              class="ͼq"
            >
              plugins
            </span>
            : 
            <span
              class="ͼ12"
            >
              [
            </span>
          </div>
          <div
            class="cm-line"
          >
                      
            <span
              class="ͼs"
            >
              viteTsConfigPaths
            </span>
            (
            <span
              class="ͼw"
            >
              {
            </span>
          </div>
          <div
            class="cm-line"
          >
                        
            <span
              class="ͼq"
            >
              root
            </span>
            : 
            <span
              class="ͼ17"
            >
              '../../../'
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                      
            <span
              class="ͼw"
            >
              }
            </span>
            )
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                    
            <span
              class="ͼ12"
            >
              ]
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
                  
            <span
              class="ͼw"
            >
              }
            </span>
            )
            <span
              class="ͼv"
            >
              ;
            </span>
          </div>
          <div
            class="cm-line"
          >
                
            <span
              class="ͼw"
            >
              }
            </span>
            <span
              class="ͼv"
            >
              ,
            </span>
          </div>
          <div
            class="cm-line"
          >
              
            <span
              class="ͼw"
            >
              }
            </span>
            <span
              class="ͼv"
            >
              ;
            </span>
          </div>
          <div
            class="cm-line"
          >
              
          </div>
          <div
            class="cm-line"
          >
              
            <span
              class="ͼ1b"
            >
              // To customize your Vite configuration you can use the viteFinal field.
            </span>
          </div>
          <div
            class="cm-line"
          >
              
            <span
              class="ͼ1b"
            >
              // Check https://storybook.js.org/docs/react/builders/vite#configuration
            </span>
          </div>
          <div
            class="cm-line"
          >
              
            <span
              class="ͼ1b"
            >
              // and https://nx.dev/packages/storybook/documents/custom-builder-configs
            </span>
          </div>
        </div>
        <div
          aria-hidden="true"
          class="cm-layer cm-layer-above cm-cursorLayer"
          style="z-index: 150; animation-duration: 1200ms;"
        />
        <div
          aria-hidden="true"
          class="cm-layer cm-selectionLayer"
          style="z-index: -2;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
